### ISSUE TITLE
ListOf AMP Arguments should support 'optional' flag

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4474]
|<img alt="faldridge's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| faldridge reported|
|-|-|
|Trac ID|trac#4474|
|Type|enhancement|
|Created|2010-05-31 22:17:40Z|

Like all other AMP Argument subclasses, t.p.amp.ListOf should allow callers to pass 'optional=True' to its __init__() method, thereby allowing the argument to be omitted from the protocol.

Attached is a patch that provides the correct behavior with the necessary update to ListOf's docstring.

As the functionality is handled by the base class (t.p.amp.Argument), and not individual subclasses, and as the functionality already has unit tests, there are no unit tests included in this patch.

Attachments:

* [listof-optional-4474.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/595/5958f8e68cd88f8b46609ff5fe0c0b8c484a157f/9152f8a13e3d5e3d39560990d96ad9c2c34cd2da.patch) (3212 bytes) - added by faldridge on 2010-06-01 18:48:05Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4474 4474
type__enhancement enhancement
reporter__faldridge faldridge
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1275344260000000 1275344260000000
changetime__1275424638000000 1275424638000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to faldridge|
|-|-|

The patch looks good.  Thanks.  A couple things though:

  1. In @foo docs, after the first line, lines should be indented 4 spaces.
  1. The feature should have unit tests.

**Comment by twisted-trac** [comment]
|<img alt="faldridge's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|faldridge set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_1):
> The patch looks good.  Thanks.  A couple things though:
> 
>   1. In @foo docs, after the first line, lines should be indented 4 spaces.
>   2. The feature should have unit tests.

1. Done.
2. Added.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Thanks, looks good, I'll apply (including a news fragment).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [29159]) Add `optional` parameter to `twisted.protocols.amp.ListOf`

Author: faldridge
Reviewer: exarkun
Fixes: [#4474](https://github.com/twisted/twisted/issues/4474)

Let `ListOf` arguments be optional in the usual way.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
